[develop] Introduce DA data preprocessing to workflow

mkavulich opened the pull request:
## DESCRIPTION OF CHANGES: 
This is a set of contributions as part of the RRFS/dev merge. The major capabilities include:

 - A new task (`get_da_obs`) that retrieves and stages observation files for the subsequent data processing tasks. Most of this logic is existing consolidated logic that has been moved out of the `process*` tasks.
 - Add workflow entries for running `get_da_obs`, `process_radarref`, `process_lightning`, and `process_bufrobs` tasks
 - Add a new `config.rrfs.yaml` file for RRFS-specific options; this will be expanded over time as more capabilities are added.
 - Add a new test (`process_obs`) to test these initial observation processing tasks. **Note that in order to run this test, the app must be built with the rrfs_utils and gsi builds, which is not part of the default build at this time.**
 - More verbose errors for `run_WE2E_tests.py`
 - Improvements to retrieve_data.py
   - Rename `--file_type` and `--external_model` arguments to `--file_fmt` and `--data_type`, respectively. These are less confusing option names since both model data and obs can be retrieved with this script.
   - Fix bug where script was not looking for files in additional locations if they exist
   - Fix bug where script failed if *any* files are missing from zip archive; this should not be a failure condition unless *all* are missing
   - Make some arguments optional if it's possible to continue without them
   - Give more helpful/verbose error messages if an invalid/unavailable data store was selected
   - Add new unit tests "test_rap_obs_from_hpss" and "test_rap_e_obs_from_hpss"

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## TESTS CONDUCTED: 
Ran fundamental tests on Hera and Jet, all passed. Ran new "process_obs" test on Hera and Jet that just tests the observation preprocessing tasks, confirmed that all expected tasks ran and completed successfully.

- [x] hera.intel
  - [x] fundamental test suite
- [x] jet.intel
  - [x] fundamental test suite
  - [x] test_retrieve_data.py unit tests

## DEPENDENCIES:
None

## DOCUMENTATION:
None

## ISSUE: 
Resolves #675 

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes do not require updates to the documentation (explain).
  - RRFS documentations will be updated when merge complete
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

## CONTRIBUTORS (optional): 
Workflow changes contributed by @christinaholtNOAA.

mkavulich committed:
First draft, doesnt work

mkavulich committed:
Emergency commit, still not working

mkavulich committed:
Mods to retrieve_data script for retrieving observations

 - "archive_internal_dir" must be an empty string for top-level archive
   files
 - Rename "file_type" argument to "file_fmt"
 - Rename "external_model" argument to more generic/correct "data_type"
 - If only some files can not be un-zipped, do not fail. We'll handle
   this more gracefully later if it needs to be an error
 - Remove default location logic for now; couldn't get it working
 - Re-add explicit disk path requirement; may re-visit later
 - For fatal error, do not imply that script continues
 - Remove debug prints

mkavulich committed:
First draft of get obs exscript. Modified data_locations to look for
"rap_e" files in addition to "rap".

mkavulich committed:
First version that makes a correct XML! Still a few kinks to work out:
specifically HPSS fails with some kind of logging error. Needs more
investigation.

This commit removes some un-needed, un-initialized variables, adds
modulefiles for the new task on each platform, and adds some unit tests
for retrieve_data.py to test new capabilities.

The new task entry has been added to parm/wflow/coldstart.yaml for
testing, but it should be removed prior to merging this PR since the
task is not integrated with the rest of the workflow yet.

mkavulich committed:
It works! Just needed to specify HPSS partition

mkavulich committed:
Start new workflow yaml "warmstart.yaml" that will be used for GSI/DA tasks going forward. Clearly the boundary conditions tasks will need to eventually change also.

mkavulich committed:
Prevent user from entering empty --tests argument to run_WE2E_tests.py

mkavulich committed:
Fix incorrect path after rebase, add alternate path for older RAP data

christinaholtNOAA committed:
Untested first pass at defining the early workflow

mkavulich committed:
Starting to link obs to tasks

 - Link obs needed for task_process_bufr
 - Put obs in $DATA/obs rather than $DATA/gsi
 - Put in logic for getting "rap_e" files at 00 and 12z cycles; I hope
we can extract this logic into the config file later with Jinja
 - Add "task_process_bufr" entry to warmstart.yaml with a dependency on
task get_da_obs
 - Add "CYCLE_TYPE" variable to config_defaults
 - Start a new config.rrfs.yaml file

mkavulich committed:
Some better design choices after putting some more thought into it

 - Set machine-specific paths in machine files, this should have been
obvious...
 - Use RRFS_NA_13km domain for RRFS
 - Re-organize Jet and Hera machine files for section consistency
 - Fix grammar error in retrieve_data.py

mkavulich committed:
Data now successfully staged for process_bufr task!

 - Move date variables to J-job level
 - Keep da_obs in same directory for community and nco mode (since
process_bufr does the same)
 - In process_bufr, look for da_obs in staging directory

mkavulich committed:
Add NLDN lightning data to get_da_obs

 - Add default directory to Jet machine file...only available there for
now
 - Add task_process_lightning to warmstart workflow

mkavulich committed:
Adding NLDN lightning to get_da_obs. Not fully tested yet, need to move to Hera. Also fix config.rrfs.yaml to better settings for test case.

mkavulich committed:
Use "else" statement in final try block for generate_FV3LAM_wflow.py to avoid duplicate error messages, search for RRFS data on disk as well as hpss

mkavulich committed:
Hot diggity, got the process_bufr and process_lightning tasks working on Hera!

mkavulich committed:
Partial incorporation of Christina's changes from #725

christinaholtNOAA committed:
CYCLE_TYPE is a run-time arg.

Removing the CYCLE_TYPE from config_defaults because it is a run-time,
cycle-dependent setting, not an experiment level one.

christinaholtNOAA committed:
Make paths and data stores configurable.

This is similar to what is done for external model files so that we can
use the machine file information about which data stores a platform has
access to.

christinaholtNOAA committed:
Remove duplicative information from machine files.

christinaholtNOAA committed:
Renaming process_bufr to process_bufrobs.

Matching new standard outlined in Issue #633

christinaholtNOAA committed:
Working prepbufr getting and processing workflow.

christinaholtNOAA committed:
This version of rrfs_beta might work.

christinaholtNOAA committed:
Turn it off since we can't use it.

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/add_get_obs_task

Had to manually revert some renaming changes in ush/test_retrieve_data.py

mkavulich committed:
Updates to fix failing unit tests

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/add_get_obs_task

mkavulich committed:
Update get_da_obs modulefiles for latest develop changes

MichaelLueken commented on ush/test_retrieve_data.py:
```diff
@@ -494,3 +496,59 @@ def test_ufs_lbcs_from_aws(self):
             path = os.path.join(tmp_dir, "*")
             files_on_disk = glob.glob(path)
             self.assertEqual(len(files_on_disk), 2)
+
```
The unit tests appear to be failing for the newly added `test_rap_obs_from_hpss` and `test_rap_e_obs_from_hpss` tests due to HPSS not being available.  These tests should pass if:

`@unittest.skipIf(os.environ.get("CI") == "true", "Skipping HPSS tests")`

is added before these two tests.

MichaelLueken reviewed (commented):
@mkavulich I haven't tried running the new `process_obs` test at this time.  I think I see why the `Python Functional Tests` is currently failing and have provided feedback in my review.

For the ufs-case-study tests, do we want to skip running them (in this PR), or move forward with @christinaholtNOAA's PR #776 to check for the existence of the files, rather than download them?

mkavulich committed:
Forgot to add CI sckip for new HPSS tests

mkavulich committed:
Remove UFS case study skip, will go with propsal from #776

mkavulich commented:
@MichaelLueken thanks for the comments; I accidentally left that skip for the UFS case study tests in, I've removed it in favor of the solution from #776. Also added the necessary HPSS skips, so the next round of tests should pass.

mkavulich commented:
@MichaelLueken thanks for the comments; I accidentally left that skip for the UFS case study tests in, I've removed it in favor of the solution from #776. Also added the necessary HPSS skips, so the next round of tests should pass.

MichaelLueken reviewed (approved):
@mkavulich Thank you very much for making the requested modifications!  I was able to successfully run the `process_obs` test on Hera, so I will now approve this work.

MichaelLueken commented:
@mkavulich Following the merging of PR #728 this morning, there is now a conflict in `ush/test_retrieve_data.py`.  Please merge the latest develop into your `feature/add_get_obs_task` branch at your earliest convenience.

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/add_get_obs_task

mkavulich committed:
Add command-line instructions for unit tests

mkavulich committed:
Change "two days ago" tests to "three days ago"...occasionally some data isnt ready by 00z two days later

mkavulich committed:
Remove references to current working directory; this causes occasional problems when running all unit tests

mkavulich commented:
@MichaelLueken I have merged in the latest changes from develop, and fixed some problems with the merged unit tests. I can not figure out why the linter is unhappy (I ran the linter on Jet for develop and got 9.74/10, so I don't know what I'm supposed to be aiming for). I have asked @christinaholtNOAA to take a look.

MichaelLueken commented:
@mkavulich It looks like the new linting GitHub Action test gave you a score of 9.98/10.  I see the following message:

```
************* Module test_python.test_calculate_cost
tests/test_python/test_calculate_cost.py:1:0: R0801: Similar lines in 2 files
==test_python.test_set_FV3nml_ens_stoch_seeds:30
==test_python.test_set_FV3nml_sfc_climo_filenames:26
        test_dir = os.path.dirname(os.path.abspath(__file__))
        USHdir = os.path.join(test_dir, "..", "..", "ush")
        PARMdir = os.path.join(USHdir, "..", "parm")

        # Create a temporary experiment directory structure
        self.tmp_dir = tempfile.TemporaryDirectory(
            dir=os.path.dirname(__file__),
            prefix="expt",
            )
        EXPTDIR = self.tmp_dir.name (duplicate-code)

-----------------------------------
```

christinaholtNOAA committed:
Allow this sort of similarity in tests.

christinaholtNOAA commented:
@mkavulich may have been the first one to check my work on-prem. I've pinned the GitHub actions to pylint v2.6, which is consistent with what's available in an old version of regional_workflow on Jet/Hera. I need to open an Issue/PR and bump that version to the pylint version available in the appropriate supported environments. This should help folks debug locally before pushing. 

I'm sorry for the inconvenience.

Edit: I opened Issue #787 to document the need for this change.

EdwardSnyder-NOAA commented:
Ran the `process_obs` WE2E test, `config.rrfs.yaml` experiment, and `test_retrieve_data.py` unit tests on Hera without any issues. All code checks out as well. The only suggestion I have is to perhaps rename the `config.rrfs.yaml` to something else, since we aren't releasing the full RRFS app, and the naming could be deceiving.

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into feature/add_get_obs_task

EdwardSnyder-NOAA commented:
Approving based on our talk yesterday that the `config.rrfs.yaml` will be renamed.

MichaelLueken commented on tests/test_python/test_retrieve_data.py:
```diff
@@ -481,15 +481,79 @@ def test_ufs_lbcs_from_aws(self):
             # Testing that there is no failure
             retrieve_data.main(args)
+
+            # Verify files exist in temp dir
+
+            path = os.path.join(tmp_dir, "*")
+            files_on_disk = glob.glob(path)
+            self.assertEqual(len(files_on_disk), 2)
```
@mkavulich - It looks like these lines are causing the Python Functional Tests to fail.  Since this test is using the new --check_file flag, no files are being downloaded, so the verification that files exist in temp dir will fail, leading to the failure in the Python Functional Tests.

MichaelLueken replied:
This has been resolved with hash [ae8ebc8](https://github.com/ufs-community/ufs-srweather-app/pull/778/commits/ae8ebc88c1f736b05231678610b53c87834b0c01).  Resolving now.

christinaholtNOAA committed:
Update wording to reflect that this is not RRFS.

christinaholtNOAA committed:
Rename the community config for da cycling.

christinaholtNOAA committed:
Attempt to fix test that only checks the URL.

christinaholtNOAA commented:
It looks like many of the major platform/compiler combos passed the Jenkins tests and Gaea's lack of Intel compiler module held up the others (best I can tell).

Can anyone help us get this PR through the pipeline given that the PR itself looks fine, but not Gaea?

natalie-perlin commented:
@christinaholtNOAA - Gaea has been updated, and new regression tests now pass with the newly updated stack for the C3 and C4 partitions, see [WM issue-1755 (1)
](https://github.com/ufs-community/ufs-weather-model/issues/1755#issuecomment-1553042387), [WM issue-1755 (2)
](https://github.com/ufs-community/ufs-weather-model/issues/1755#issuecomment-1553253952)

At this point it is a matter of updating the ufs-weather-model repository pointing to the new software stack.

christinaholtNOAA commented:
Okay, but what does that mean for this PR?

natalie-perlin commented:
@christinaholtNOAA - Could Gaea possibly be added to the testing process? (still pending updates to the stack location)

christinaholtNOAA commented:
@natalie-perlin Are you suggesting that we update the version of the weather model in this PR to make the Gaea tests pass?

That is way outside the scope of the work in this PR, and it would be great if we could find a solution that lets us merge the current PR as-is since the failures are due to factors outside our control and we have the approvals.

We're also only adding new capabilities here, not modifying old ones.

natalie-perlin commented:
@christinaholtNOAA - Oh, the question was whether any changes in machine file for Gaea are expected, but not implemented due to issues following the upgrade. If no changes are expected for Gaea, and it only holds the PR from being fully passed, then no objections!

MichaelLueken commented:
On Cheyenne GNU, the `nco_grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15_thompson_mynn_lam3km` test failed intermittently in run_post.  This is an expected failure.

The Gaea tests will fail until PR #799 is merged.

On Hera GNU, the `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS` test failed in the get_extrn_ics/lbcs tasks.  Since this test has been added, it fails intermittently with this same issue.

On Hera Intel, the `grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` test is failing in the run_fcst task.  This is an expected failure.

Since all of the failures are expected, I will go ahead and merge this work now.
